File: ofd/page.go
package ofd

import "encoding/xml"

// Page 页面结构
type Page struct {
	XMLName   xml.Name   `xml:"Page"`
	Area      PageArea   `xml:"Area"`
	Content   Content    `xml:"Content"`
	ContentNS Content    `xml:"http://www.ofdspec.org/2016 Content"`
	Template  []Template `xml:"Template"`
	// 直接包含 Layer（某些 OFD 文档可能没有 Content 包装）
	Layer []Layer `xml:"Layer"`
}

// Template 模板引用
type Template struct {
	TemplateID string `xml:"TemplateID,attr"`
	ZOrder     string `xml:"ZOrder,attr"`
}

// Content 页面内容
type Content struct {
	Layer []Layer `xml:"Layer"`
}

// Layer 图层
type Layer struct {
	ID           string        `xml:"ID,attr"`
	Type         string        `xml:"Type,attr"`
	DrawParam    string        `xml:"DrawParam,attr"`
	TextObjects  []TextObject  `xml:"TextObject"`
	PathObjects  []PathObject  `xml:"PathObject"`
	ImageObjects []ImageObject `xml:"ImageObject"`
}

// TextObject 文本对象
type TextObject struct {
	ID            string        `xml:"ID,attr"`
	Boundary      string        `xml:"Boundary,attr"`
	Font          string        `xml:"Font,attr"`
	Size          float64       `xml:"Size,attr"`
	HScale        float64       `xml:"HScale,attr"`
	CharSpace     float64       `xml:"CharSpace,attr"`
	ReadDirection int           `xml:"ReadDirection,attr"`
	Weight        int           `xml:"Weight,attr"`
	Italic        bool          `xml:"Italic,attr"`
	Stroke        bool          `xml:"Stroke,attr"`
	Fill          bool          `xml:"Fill,attr"`
	Visible       *bool         `xml:"Visible,attr"`
	LineWidth     float64       `xml:"LineWidth,attr"`
	CTM           string        `xml:"CTM,attr"`
	FillColor     *Color        `xml:"FillColor"`
	StrokeColor   *Color        `xml:"StrokeColor"`
	TextCode      []TextCode    `xml:"TextCode"`
	CGTransform   []CGTransform `xml:"CGTransform"`
}

// TextCode 文本内容
type TextCode struct {
	X       float64 `xml:"X,attr"`
	Y       float64 `xml:"Y,attr"`
	DeltaX  string  `xml:"DeltaX,attr"`
	DeltaY  string  `xml:"DeltaY,attr"`
	Content string  `xml:",chardata"`
}

// CGTransform 变换矩阵
type CGTransform struct {
	CodePosition int    `xml:"CodePosition,attr"`
	CodeCount    int    `xml:"CodeCount,attr"`
	GlyphCount   int    `xml:"GlyphCount,attr"`
	GlyphsAttr   string `xml:"Glyphs,attr"`   // Glyphs 作为属性
	GlyphsElem   string `xml:"Glyphs"`        // Glyphs 作为子元素
}

// GetGlyphs 获取 Glyphs 值（兼容属性和子元素两种格式）
func (c *CGTransform) GetGlyphs() string {
	if c.GlyphsAttr != "" {
		return c.GlyphsAttr
	}
	return c.GlyphsElem
}

// PathObject 路径对象
type PathObject struct {
	ID              string       `xml:"ID,attr"`
	Boundary        string       `xml:"Boundary,attr"`
	CTM             string       `xml:"CTM,attr"`
	LineWidth       float64      `xml:"LineWidth,attr"`
	Join            string       `xml:"Join,attr"`        // 线条连接样式：Miter, Round, Bevel
	Cap             string       `xml:"Cap,attr"`         // 线条端点样式：Butt, Round, Square
	Rule            string       `xml:"Rule,attr"`        // 填充规则：NonZero, Even-Odd
	Stroke          bool         `xml:"Stroke,attr"`
	Fill            bool         `xml:"Fill,attr"`
	Visible         *bool        `xml:"Visible,attr"`
	FillColor       *ColorOrShd  `xml:"FillColor"`
	StrokeColor     *Color       `xml:"StrokeColor"`
	AbbreviatedData string       `xml:"AbbreviatedData"`
}

// ImageObject 图像对象
type ImageObject struct {
	ID         string  `xml:"ID,attr"`
	Boundary   string  `xml:"Boundary,attr"`
	ResourceID string  `xml:"ResourceID,attr"`
	CTM        string  `xml:"CTM,attr"`
	Alpha      int     `xml:"Alpha,attr"`
	Visible    *bool   `xml:"Visible,attr"`
}

// Color 颜色
type Color struct {
	Value      string `xml:"Value,attr"`
	ColorSpace string `xml:"ColorSpace,attr"`
}

// ColorOrShd 颜色或渐变或图案
type ColorOrShd struct {
	Value      string     `xml:"Value,attr"`
	ColorSpace string     `xml:"ColorSpace,attr"`
	AxialShd   *AxialShd  `xml:"AxialShd"`
	RadialShd  *RadialShd `xml:"RadialShd"`
	Pattern    *Pattern   `xml:"Pattern"`
}

// Pattern 图案填充
type Pattern struct {
	Width      float64     `xml:"Width,attr"`      // 单元格内容宽度 (mm)
	Height     float64     `xml:"Height,attr"`     // 单元格内容高度 (mm)
	XStep      float64     `xml:"XStep,attr"`      // 水平平铺步长 (mm)
	YStep      float64     `xml:"YStep,attr"`      // 垂直平铺步长 (mm)
	RelativeTo string      `xml:"RelativeTo,attr"` // "Page" 或 "Object"
	CTM        string      `xml:"CTM,attr"`        // 变换矩阵
	CellContent CellContent `xml:"CellContent"`    // 单元格内容
}

// CellContent Pattern 单元格内容
type CellContent struct {
	ImageObjects []ImageObject `xml:"ImageObject"`
	PathObjects  []PathObject  `xml:"PathObject"`
	TextObjects  []TextObject  `xml:"TextObject"`
}

// AxialShd 轴向渐变（线性渐变）
type AxialShd struct {
	StartPoint string    `xml:"StartPoint,attr"`
	EndPoint   string    `xml:"EndPoint,attr"`
	Extend     string    `xml:"Extend,attr"`
	Segment    []Segment `xml:"Segment"`
}

// RadialShd 径向渐变
type RadialShd struct {
	StartPoint string    `xml:"StartPoint,attr"`
	EndPoint   string    `xml:"EndPoint,attr"`
	StartRadius float64  `xml:"StartRadius,attr"`
	EndRadius   float64  `xml:"EndRadius,attr"`
	Extend      string   `xml:"Extend,attr"`
	Segment     []Segment `xml:"Segment"`
}

// Segment 渐变段
type Segment struct {
	Position float64 `xml:"Position,attr"`
	Color    Color   `xml:"Color"`
}

// Res 资源文件
type Res struct {
	XMLName     xml.Name     `xml:"Res"`
	BaseLoc     string       `xml:"BaseLoc,attr"`
	Fonts       []Font       `xml:"Fonts>Font"`
	MultiMedias []MultiMedia `xml:"MultiMedias>MultiMedia"`
	ColorSpaces []ColorSpace `xml:"ColorSpaces>ColorSpace"`
}

// Font 字体
type Font struct {
	ID         string `xml:"ID,attr"`
	FontName   string `xml:"FontName,attr"`
	FamilyName string `xml:"FamilyName,attr"`
	FontFile   string `xml:"FontFile"` // 子元素，不是属性
}

// MultiMedia 多媒体资源
type MultiMedia struct {
	ID        string `xml:"ID,attr"`
	Type      string `xml:"Type,attr"`
	MediaFile string `xml:"MediaFile"`
}

// ColorSpace 颜色空间
type ColorSpace struct {
	ID   string `xml:"ID,attr"`
	Type string `xml:"Type,attr"`
}

// Signatures 签章结构
type Signatures struct {
	XMLName   xml.Name    `xml:"Signatures"`
	Signature []Signature `xml:"Signature"`
}

// Signature 签章
type Signature struct {
	ID      string `xml:"ID,attr"`
	Type    string `xml:"Type,attr"`
	BaseLoc string `xml:"BaseLoc,attr"` // BaseLoc 是属性，不是子元素
}

// SignatureXML 签章XML - 支持多种格式和命名空间
type SignatureXML struct {
	XMLName     xml.Name   `xml:"Signature"`
	SignedInfo  SignedInfo `xml:"SignedInfo"`
	SignedValue string     `xml:"SignedValue"`
}

// SignatureXMLNS 带命名空间的签章XML
type SignatureXMLNS struct {
	XMLName     xml.Name   `xml:"http://www.ofdspec.org/2016 Signature"`
	SignedInfo  SignedInfo `xml:"http://www.ofdspec.org/2016 SignedInfo"`
	SignedValue string     `xml:"http://www.ofdspec.org/2016 SignedValue"`
}

// SignedInfo 签章信息
type SignedInfo struct {
	StampAnnot   []StampAnnot `xml:"StampAnnot"`
	Seal         SealRef      `xml:"Seal"`
	References   References   `xml:"References"`
	Provider     string       `xml:"Provider"`
	SignatureMethod string    `xml:"SignatureMethod"`
	// 带命名空间的版本
	StampAnnotNS []StampAnnot `xml:"http://www.ofdspec.org/2016 StampAnnot"`
	SealNS       SealRef      `xml:"http://www.ofdspec.org/2016 Seal"`
	// ofd: 前缀版本
	StampAnnotOFD []StampAnnot `xml:"ofd StampAnnot"`
}

// SealRef 印章引用
type SealRef struct {
	BaseLocAttr string `xml:"BaseLoc,attr"` // BaseLoc 作为属性
	BaseLocElem string `xml:"BaseLoc"`      // BaseLoc 作为子元素
}

// GetBaseLoc 获取 BaseLoc（兼容属性和子元素两种格式）
func (s *SealRef) GetBaseLoc() string {
	if s.BaseLocAttr != "" {
		return s.BaseLocAttr
	}
	return s.BaseLocElem
}

// References 引用
type References struct {
	Reference []Reference `xml:"Reference"`
}

// Reference 引用
type Reference struct {
	FileRef string `xml:"FileRef,attr"`
}

// StampAnnot 印章注释
type StampAnnot struct {
	ID       string `xml:"ID,attr"`
	PageRef  string `xml:"PageRef,attr"`
	Boundary string `xml:"Boundary,attr"`
}

// SealXML 印章数据
type SealXML struct {
	XMLName     xml.Name `xml:"Seal"`
	SealID      string   `xml:"SealID"`
	Type        string   `xml:"Type"`
	Picture     Picture  `xml:"Picture"`
	SealData    string   `xml:"SealData"`
	PictureSES  PictureSES `xml:"ofd:Picture"`
}

// PictureSES SES格式图片
type PictureSES struct {
	Type   string `xml:"Type,attr"`
	Width  float64 `xml:"Width,attr"`
	Height float64 `xml:"Height,attr"`
	Data   string `xml:",chardata"`
}

// Picture 图片
type Picture struct {
	Type   string `xml:"Type,attr"`
	Width  string `xml:"Width,attr"`
	Height string `xml:"Height,attr"`
	Data   string `xml:",chardata"`
}


// Annots 注释结构
type Annots struct {
	XMLName xml.Name `xml:"PageAnnot"`
	Annot   []Annot  `xml:"Annot"`
}

// Annot 注释
type Annot struct {
	ID         string      `xml:"ID,attr"`
	Type       string      `xml:"Type,attr"`
	Creator    string      `xml:"Creator,attr"`
	LastModify string      `xml:"LastModify,attr"`
	Subtype    string      `xml:"Subtype,attr"`
	Visible    bool        `xml:"Visible,attr"`
	Print      bool        `xml:"Print,attr"`
	NoZoom     bool        `xml:"NoZoom,attr"`
	NoRotate   bool        `xml:"NoRotate,attr"`
	Appearance Appearance  `xml:"Appearance"`
}

// Appearance 外观
type Appearance struct {
	Boundary   string      `xml:"Boundary,attr"`
	PageBlocks []PageBlock `xml:"PageBlock"`
}

// PageBlock 页面块
type PageBlock struct {
	ID           string        `xml:"ID,attr"`
	ImageObjects []ImageObject `xml:"ImageObject"`
	PathObjects  []PathObject  `xml:"PathObject"`
	TextObjects  []TextObject  `xml:"TextObject"`
}

// PageAnnot 页面注释（用于解析注释文件）
type PageAnnot struct {
	XMLName xml.Name       `xml:"PageAnnot"`
	Annots  []AnnotElement `xml:"Annot"`
}

// AnnotElement 注释元素
type AnnotElement struct {
	ID         string            `xml:"ID,attr"`
	Type       string            `xml:"Type,attr"`
	Creator    string            `xml:"Creator,attr"`
	Subtype    string            `xml:"Subtype,attr"`
	Appearance AnnotAppearance   `xml:"Appearance"`
}

// AnnotAppearance 注释外观
type AnnotAppearance struct {
	Boundary     string            `xml:"Boundary,attr"`
	PageBlocks   []AnnotPageBlock  `xml:"PageBlock"`
	// 直接包含的对象（某些 OFD 文档 Appearance 下直接放对象，不包 PageBlock）
	ImageObjects []ImageObject     `xml:"ImageObject"`
	PathObjects  []PathObject      `xml:"PathObject"`
	TextObjects  []TextObject      `xml:"TextObject"`
}

// AnnotPageBlock 注释页面块
type AnnotPageBlock struct {
	ID           string        `xml:"ID,attr"`
	ImageObjects []ImageObject `xml:"ImageObject"`
	PathObjects  []PathObject  `xml:"PathObject"`
	TextObjects  []TextObject  `xml:"TextObject"`
}

// AnnotationsFile Annotations.xml 索引文件
type AnnotationsFile struct {
	XMLName xml.Name        `xml:"Annotations"`
	Pages   []AnnotPage     `xml:"Page"`
}

// AnnotPage 注释索引中的页面条目
type AnnotPage struct {
	PageID  string `xml:"PageID,attr"`
	FileLoc string `xml:"FileLoc"`
}

File: ofd-rust/src/page.rs
//! 页面结构定义

use serde::{Deserialize, Serialize};

/// 页面结构
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
#[serde(rename = "Page")]
pub struct Page {
    #[serde(rename = "Area", default)]
    pub area: PageAreaDef,
    #[serde(rename = "Content", default)]
    pub content: Content,
    #[serde(rename = "Layer", default)]
    pub layer: Vec<Layer>,
    #[serde(rename = "Template", default)]
    pub template: Vec<Template>,
}

/// 页面区域定义
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct PageAreaDef {
    #[serde(rename = "PhysicalBox", default)]
    pub physical_box: String,
    #[serde(rename = "ApplicationBox", default)]
    pub application_box: String,
    #[serde(rename = "ContentBox", default)]
    pub content_box: String,
}

/// 模板引用
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Template {
    #[serde(rename = "@TemplateID", default)]
    pub template_id: String,
    #[serde(rename = "@ZOrder", default)]
    pub z_order: String,
}

/// 页面内容
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Content {
    #[serde(rename = "Layer", default)]
    pub layer: Vec<Layer>,
}

/// 图层
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Layer {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Type", default)]
    pub layer_type: String,
    #[serde(rename = "@DrawParam", default)]
    pub draw_param: String,
    #[serde(rename = "$value", default)]
    pub objects: Vec<LayerObject>,
}

/// 图层对象（可以是文本、路径或图片）
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "PascalCase")]
pub enum LayerObject {
    TextObject(TextObject),
    PathObject(PathObject),
    ImageObject(ImageObject),
    CompositeObject(CompositeObject),
}

/// 复合对象引用
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CompositeObject {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Boundary", default)]
    pub boundary: String,
    #[serde(rename = "@ResourceID", default)]
    pub resource_id: String,
    #[serde(rename = "@CTM", default)]
    pub ctm: String,
}

impl Layer {
    pub fn text_objects(&self) -> impl Iterator<Item = &TextObject> {
        self.objects.iter().filter_map(|o| match o {
            LayerObject::TextObject(t) => Some(t),
            _ => None,
        })
    }

    pub fn path_objects(&self) -> impl Iterator<Item = &PathObject> {
        self.objects.iter().filter_map(|o| match o {
            LayerObject::PathObject(p) => Some(p),
            _ => None,
        })
    }

    pub fn image_objects(&self) -> impl Iterator<Item = &ImageObject> {
        self.objects.iter().filter_map(|o| match o {
            LayerObject::ImageObject(i) => Some(i),
            _ => None,
        })
    }

    pub fn composite_objects(&self) -> impl Iterator<Item = &CompositeObject> {
        self.objects.iter().filter_map(|o| match o {
            LayerObject::CompositeObject(c) => Some(c),
            _ => None,
        })
    }
}

/// 文本对象
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TextObject {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Boundary", default)]
    pub boundary: String,
    #[serde(rename = "@Font", default)]
    pub font: String,
    #[serde(rename = "@Size", default)]
    pub size: f64,
    #[serde(rename = "@HScale", default)]
    pub h_scale: f64,
    #[serde(rename = "@Weight", default)]
    pub weight: i32,
    #[serde(rename = "@Italic", default)]
    pub italic: bool,
    #[serde(rename = "@Stroke", default)]
    pub stroke: bool,
    #[serde(rename = "@Fill", default = "default_true")]
    pub fill: bool,
    #[serde(rename = "@LineWidth", default = "default_line_width")]
    pub line_width: f64,
    #[serde(rename = "@CTM", default)]
    pub ctm: String,
    #[serde(rename = "@DrawParam", default)]
    pub draw_param: String,
    #[serde(rename = "@Alpha", default = "default_alpha")]
    pub alpha: i32,
    #[serde(rename = "FillColor", default)]
    pub fill_color: Option<Color>,
    #[serde(rename = "StrokeColor", default)]
    pub stroke_color: Option<Color>,
    #[serde(rename = "TextCode", default)]
    pub text_code: Vec<TextCode>,
    #[serde(rename = "CGTransform", default)]
    pub cg_transform: Vec<CGTransform>,
}

impl Default for TextObject {
    fn default() -> Self {
        TextObject {
            id: String::new(),
            boundary: String::new(),
            font: String::new(),
            size: 0.0,
            h_scale: 0.0,
            weight: 0,
            italic: false,
            stroke: false,
            fill: true,
            line_width: 0.353,
            ctm: String::new(),
            draw_param: String::new(),
            alpha: 255,
            fill_color: None,
            stroke_color: None,
            text_code: Vec::new(),
            cg_transform: Vec::new(),
        }
    }
}

fn default_true() -> bool { true }
fn default_line_width() -> f64 { 0.353 }
fn default_alpha() -> i32 { 255 }

/// 文本内容
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct TextCode {
    #[serde(rename = "@X", default)]
    pub x: f64,
    #[serde(rename = "@Y", default)]
    pub y: f64,
    #[serde(rename = "@DeltaX", default)]
    pub delta_x: String,
    #[serde(rename = "@DeltaY", default)]
    pub delta_y: String,
    #[serde(rename = "$text", default)]
    pub content: String,
}

/// 字形变换
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CGTransform {
    #[serde(rename = "@CodePosition", default)]
    pub code_position: i32,
    #[serde(rename = "@CodeCount", default)]
    pub code_count: i32,
    #[serde(rename = "@GlyphCount", default)]
    pub glyph_count: i32,
    #[serde(rename = "Glyphs", default)]
    pub glyphs: String,
}

/// 路径对象
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct PathObject {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Boundary", default)]
    pub boundary: String,
    #[serde(rename = "@CTM", default)]
    pub ctm: String,
    #[serde(rename = "@LineWidth", default)]
    pub line_width: f64,
    #[serde(rename = "@Join", default)]
    pub join: String,
    #[serde(rename = "@Cap", default)]
    pub cap: String,
    #[serde(rename = "@Stroke", default = "default_true")]
    pub stroke: bool,
    #[serde(rename = "@Fill", default)]
    pub fill: bool,
    #[serde(rename = "@DrawParam", default)]
    pub draw_param: String,
    #[serde(rename = "FillColor", default)]
    pub fill_color: Option<ColorOrShd>,
    #[serde(rename = "StrokeColor", default)]
    pub stroke_color: Option<Color>,
    #[serde(rename = "AbbreviatedData", default)]
    pub abbreviated_data: String,
}

/// 图像对象
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct ImageObject {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Boundary", default)]
    pub boundary: String,
    #[serde(rename = "@ResourceID", default)]
    pub resource_id: String,
    #[serde(rename = "@CTM", default)]
    pub ctm: String,
    #[serde(rename = "@Alpha", default = "default_alpha")]
    pub alpha: i32,
}

/// 颜色
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Color {
    #[serde(rename = "@Value", default)]
    pub value: String,
    #[serde(rename = "@ColorSpace", default)]
    pub color_space: String,
}

/// 颜色或渐变
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct ColorOrShd {
    #[serde(rename = "@Value", default)]
    pub value: String,
    #[serde(rename = "AxialShd", default)]
    pub axial_shd: Option<AxialShd>,
    #[serde(rename = "RadialShd", default)]
    pub radial_shd: Option<RadialShd>,
    #[serde(rename = "Pattern", default)]
    pub pattern: Option<Pattern>,
}

/// 轴向渐变
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct AxialShd {
    #[serde(rename = "@StartPoint", default)]
    pub start_point: String,
    #[serde(rename = "@EndPoint", default)]
    pub end_point: String,
    #[serde(rename = "Segment", default)]
    pub segment: Vec<Segment>,
}

/// 径向渐变
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct RadialShd {
    #[serde(rename = "@StartPoint", default)]
    pub start_point: String,
    #[serde(rename = "@EndPoint", default)]
    pub end_point: String,
    #[serde(rename = "@StartRadius", default)]
    pub start_radius: f64,
    #[serde(rename = "@EndRadius", default)]
    pub end_radius: f64,
    #[serde(rename = "Segment", default)]
    pub segment: Vec<Segment>,
}

/// 渐变段
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Segment {
    #[serde(rename = "@Position", default)]
    pub position: f64,
    #[serde(rename = "Color", default)]
    pub color: Color,
}

/// 资源文件
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
#[serde(rename = "Res")]
pub struct Res {
    #[serde(rename = "@BaseLoc", default)]
    pub base_loc: String,
    #[serde(rename = "Fonts", default)]
    pub fonts: Option<Fonts>,
    #[serde(rename = "MultiMedias", default)]
    pub multi_medias: Option<MultiMedias>,
    #[serde(rename = "CompositeGraphicUnits", default)]
    pub composite_graphic_units: Option<CompositeGraphicUnits>,
    /// DrawParam 直接作为 Res 的子元素（无 DrawParams 包装）
    #[serde(rename = "DrawParam", default)]
    pub draw_params: Vec<DrawParam>,
    /// DrawParams 包装元素（部分 OFD 文件使用）
    #[serde(rename = "DrawParams", default)]
    pub draw_params_wrapped: Option<DrawParamsWrapper>,
}

/// DrawParams 包装元素
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct DrawParamsWrapper {
    #[serde(rename = "DrawParam", default)]
    pub draw_param: Vec<DrawParam>,
}

/// 绘制参数（定义默认描边/填充样式）
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct DrawParam {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Relative", default)]
    pub relative: String,
    #[serde(rename = "@LineWidth", default)]
    pub line_width: f64,
    #[serde(rename = "@Join", default)]
    pub join: String,
    #[serde(rename = "@Cap", default)]
    pub cap: String,
    #[serde(rename = "FillColor", default)]
    pub fill_color: Option<Color>,
    #[serde(rename = "StrokeColor", default)]
    pub stroke_color: Option<Color>,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CompositeGraphicUnits {
    #[serde(rename = "CompositeGraphicUnit", default)]
    pub units: Vec<CompositeGraphicUnit>,
}

/// 复合图元定义（如印章矢量图形）
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CompositeGraphicUnit {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Width", default)]
    pub width: f64,
    #[serde(rename = "@Height", default)]
    pub height: f64,
    #[serde(rename = "Content", default)]
    pub content: CompositeContent,
}

/// 复合图元内容
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CompositeContent {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Type", default)]
    pub content_type: String,
    #[serde(rename = "PageBlock", default)]
    pub page_block: Option<CompositePageBlock>,
}

/// 复合图元页面块
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CompositePageBlock {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "$value", default)]
    pub objects: Vec<LayerObject>,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Fonts {
    #[serde(rename = "Font", default)]
    pub font: Vec<Font>,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct MultiMedias {
    #[serde(rename = "MultiMedia", default)]
    pub multi_media: Vec<MultiMedia>,
}

/// 字体
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Font {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@FontName", default)]
    pub font_name: String,
    #[serde(rename = "@FamilyName", default)]
    pub family_name: String,
    #[serde(rename = "FontFile", default)]
    pub font_file: String,
}

/// 多媒体资源
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct MultiMedia {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Type", default)]
    pub media_type: String,
    #[serde(rename = "MediaFile", default)]
    pub media_file: String,
}

/// 签章结构
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
#[serde(rename = "Signatures")]
pub struct Signatures {
    #[serde(rename = "Signature", default)]
    pub signature: Vec<Signature>,
}

/// 签章
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Signature {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Type", default)]
    pub sig_type: String,
    #[serde(rename = "@BaseLoc", default)]
    pub base_loc: String,
}

/// 签章XML
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
#[serde(rename = "Signature")]
pub struct SignatureXML {
    #[serde(rename = "SignedInfo", default)]
    pub signed_info: SignedInfo,
    #[serde(rename = "SignedValue", default)]
    pub signed_value: String,
}

/// 签章信息
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct SignedInfo {
    #[serde(rename = "StampAnnot", default)]
    pub stamp_annot: Vec<StampAnnot>,
    #[serde(rename = "Provider", default)]
    pub provider: String,
    #[serde(rename = "SignatureMethod", default)]
    pub signature_method: String,
}

/// 印章注释
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct StampAnnot {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@PageRef", default)]
    pub page_ref: String,
    #[serde(rename = "@Boundary", default)]
    pub boundary: String,
}

/// Pattern 图案填充
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct Pattern {
    #[serde(rename = "@Width", default)]
    pub width: f64,
    #[serde(rename = "@Height", default)]
    pub height: f64,
    #[serde(rename = "@XStep", default)]
    pub x_step: f64,
    #[serde(rename = "@YStep", default)]
    pub y_step: f64,
    #[serde(rename = "@RelativeTo", default)]
    pub relative_to: String,
    #[serde(rename = "@CTM", default)]
    pub ctm: String,
    #[serde(rename = "CellContent", default)]
    pub cell_content: CellContent,
}

/// CellContent Pattern 单元格内容
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CellContent {
    #[serde(rename = "ImageObject", default)]
    pub image_objects: Vec<ImageObject>,
    #[serde(rename = "PathObject", default)]
    pub path_objects: Vec<PathObject>,
    #[serde(rename = "TextObject", default)]
    pub text_objects: Vec<TextObject>,
}

/// 页面注释列表
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
#[serde(rename = "PageAnnot")]
pub struct PageAnnot {
    #[serde(rename = "Annot", default)]
    pub annots: Vec<AnnotElement>,
}

/// 注释元素
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct AnnotElement {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "@Type", default)]
    pub annot_type: String,
    #[serde(rename = "@Creator", default)]
    pub creator: String,
    #[serde(rename = "@Subtype", default)]
    pub subtype: String,
    #[serde(rename = "Appearance", default)]
    pub appearance: AnnotAppearance,
}

/// 注释外观
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct AnnotAppearance {
    #[serde(rename = "@Boundary", default)]
    pub boundary: String,
    #[serde(rename = "PageBlock", default)]
    pub page_blocks: Vec<AnnotPageBlock>,
    /// 直接嵌在 Appearance 下的对象（无 PageBlock 包裹）
    #[serde(rename = "TextObject", default)]
    pub text_objects: Vec<TextObject>,
    #[serde(rename = "PathObject", default)]
    pub path_objects: Vec<PathObject>,
    #[serde(rename = "ImageObject", default)]
    pub image_objects: Vec<ImageObject>,
}

/// 注释页面块
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct AnnotPageBlock {
    #[serde(rename = "@ID", default)]
    pub id: String,
    #[serde(rename = "$value", default)]
    pub objects: Vec<LayerObject>,
}

/// 印章数据 (Seal.xml)
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
#[serde(rename = "Seal")]
pub struct SealXML {
    #[serde(rename = "SealID", default)]
    pub seal_id: String,
    #[serde(rename = "Type", default)]
    pub seal_type: String,
    #[serde(rename = "Picture", default)]
    pub picture: SealPicture,
    #[serde(rename = "SealData", default)]
    pub seal_data: String,
}

/// 印章图片
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct SealPicture {
    #[serde(rename = "@Type", default)]
    pub picture_type: String,
    #[serde(rename = "@Width", default)]
    pub width: f64,
    #[serde(rename = "@Height", default)]
    pub height: f64,
    #[serde(rename = "$text", default)]
    pub data: String,
}

